### ISSUE TITLE
Standard Maybe and Result monads?

### ISSUE THREAD

**Issue by cgarciae** [evhub/coconut#252]
Coconut being a language, it would be nice if it included some standard monadic types such as Maybe and Result/Either. If not I might create such a library. What do you think?

**Comment by evhub** [comment]
@cgarciae That would definitely be doable. The tests right now include this definition of Maybe:
```
data Nothing():
    def __call__(self, *args):
        return Nothing()
    def __eq__(self, other):
        match Nothing() in other:
            return True
        else:
            return False
data Just(item):
    def __call__(self, *args):
        return Just <| reduce((|>), args, self.item)
    def __eq__(self, other):
        match Just(=self.item) in other:
            return True
        else:
            return False
Maybe = (Nothing, Just)

assert Just(5) <| square <| plus1 == Just(26)
assert Nothing() <| square <| plus1 == Nothing()
```
However, I think [`fn.py`](https://github.com/kachayev/fn.py) can already do monads pretty well, though I haven't really used it so I wouldn't known.

**Comment by cgarciae** [comment]
I think I'd prefer a Coconut version based a 'data' because of the ability to use pattern matching, regardless of what fn.py might offer.
I like your implementation, using the __call__ method as 'fmap' makes things very readable.

**Comment by evhub** [comment]
@cgarciae Well, the other option is that the newest version did add an `fmap` function; take a look at: <http://coconut.readthedocs.io/en/master/DOCS.html#fmap>

**Comment by cgarciae** [comment]
Did notice. I use it this week :)

**Comment by Nexus6** [comment]
You might want to check out the goodies in my Glioma library.  It has Option/Some/Nothing classes  inspired by the Scala stdlib so maybe not as cool as what you're discussing here but I've found it useful. It was written in Python originally but I ported it to Coconut.
https://github.com/Nexus6/Glioma
pip install glioma

Note that Coconut's pattern-matching and destructuring do not work with these classes, those would be nice enhancements to the library.

It had crossed my mind that it might be cool for Coconut to have its own functionally-flavored standard class library that might contain immutable collections and other useful things like Maybe and Either.  Not sure if that might not be outside the scope of the project though and maybe OO programming shouldn't be encouraged.  :-)
